File: practice/python/patterns.py
# ============================================================
# Practice: Patterns
# Summary: right-angled triangle, centered pyramid.
# Run:     python practice/python/patterns.py
# ============================================================

# ----- 1. Right-angled triangle of stars -----
def right_triangle(n):
    lines = []
    for i in range(1, n + 1):
        lines.append("*" * i)
    return lines


# ----- 2. Centered pyramid of stars (height h) -----
def pyramid(h):
    lines = []
    for i in range(1, h + 1):
        spaces = " " * (h - i)
        stars = "*" * (2 * i - 1)
        lines.append(spaces + stars)
    return lines


if __name__ == "__main__":
    print("Right triangle (N=5):")
    for line in right_triangle(5):
        print(line)
    # *
    # **
    # ***
    # ****
    # *****

    print("Pyramid (height=5):")
    for line in pyramid(5):
        print(line)
    #     *
    #    ***
    #   *****
    #  *******
    # *********

File: practice/cpp/patterns.cpp
// ============================================================
// Practice: Patterns
// Summary: right-angled triangle, centered pyramid.
// Run:     g++ -std=c++17 practice/cpp/patterns.cpp -o practice/cpp/patterns && practice/cpp/patterns
// ============================================================

#include <iostream>
#include <string>

// ----- 1. Right-angled triangle of stars -----
void rightTriangle(int n) {
    for (int i = 1; i <= n; i++)
        std::cout << std::string(i, '*') << std::endl;
}

// ----- 2. Centered pyramid of stars (height h) -----
void pyramid(int h) {
    for (int i = 1; i <= h; i++) {
        std::string spaces(h - i, ' ');
        std::string stars(2 * i - 1, '*');
        std::cout << spaces << stars << std::endl;
    }
}

int main() {
    std::cout << "Right triangle (N=5):" << std::endl;
    rightTriangle(5);
    // *
    // **
    // ***
    // ****
    // *****

    std::cout << "Pyramid (height=5):" << std::endl;
    pyramid(5);
    //     *
    //    ***
    //   *****
    //  *******
    // *********
    return 0;
}
